reference-implementation/lib/transform-stream.js

```diff
@@ -116,39 +116,25 @@ function TransformStreamErrorInternal(transformStream, e) {
   if (transformStream._readableClosed === false) {
     transformStream._readableController.error(e);
   }
-
-  transformStream._chunk = undefined;
 }
 
-function TransformStreamTransformIfNeeded(transformStream) {
-  // console.log('TransformStreamTransformIfNeeded()');
-
-  if (transformStream._chunkPending === false) {
-    return;
-  }
-
-  assert(transformStream._resolveWrite !== undefined);
+function TransformStreamTransform(transformStream, chunk) {
+  // console.log('TransformStreamTransform()');
 
-  if (transformStream._transforming === true) {
-    return;
-  }
-
-  if (transformStream._readableBackpressure === true) {
-    return;
-  }
+  assert(transformStream._errored === false);
+  assert(transformStream._transforming === false);
+  assert(transformStream._readableBackpressure === false);
 
   transformStream._transforming = true;
 
-  const chunk = transformStream._chunk;
-  transformStream._chunkPending = false;
-  transformStream._chunk = undefined;
-
   const controller = transformStream._transformStreamController;
   const transformPromise = PromiseInvokeOrNoop(transformStream._transformer,
                              'transform', [chunk, controller]);
 
-  transformPromise.then(() => TransformStreamResolveWrite(transformStream),
-                     e => TransformStreamErrorIfNeeded(transformStream, e));
+  return transformPromise.then(() => {
+    transformStream._transforming = false;
+  }).then(() => TransformStreamReadyPromise(transformStream),
+          e => TransformStreamErrorIfNeeded(transformStream, e));
```

Comment:
I don't think this will be easy to read when translated into an algorithm in the spec.

How about

```javascript
return transformPromise.then(
    () => {
      transformStream._transforming = false;
      TransformStreamReadyPromise(transformStream);
    },
    e => TransformStreamErrorIfNeeded(transformStream, e));
```